CONTRIBUTING.md

```diff
@@ -31,10 +31,135 @@ Please be aware of the following things when filing bug reports.
 
 If the bug report is missing this information then it'll take us longer to fix the issue. We will probably ask for clarification, and barring that if no response was given then the issue will be closed.
 
+## Writing a Pull Request
+Submitting a pull request is fairly simple, just make sure it focuses on a single aspect and doesn't manage to have scope creep and it's probably good to go.
 
-## Submitting a Pull Request
+Most attributes and data structures are broken up in to a file for each related class. For example, `disnake.Guild` is defined in [disnake/guild.py](disnake/guild.py), and `disnake.GuildPreview` is defined in [disnake/guild_preview.py](disnake/guild_preview.py). For example, writing a new feature to `disnake.Guild` would go in [disnake/guild.py](disnake/guild.py), as part of the `disnake.Guild` class.
 
-Submitting a pull request is fairly simple, just make sure it focuses on a single aspect and doesn't manage to have scope creep and it's probably good to go. It would be incredibly lovely if the style is consistent to that found in the project. This project follows PEP-8 guidelines (mostly) with a column limit of 100 characters.
+### Adding a new API Feature
+
+However, adding a new feature that interfaces with the API requires also updating the [disnake/types](disnake/types) directory. We request that when making or receiving payloads from the API, they are typed and the typehints are used on the functions that are processing said data. As an example, we'll take a look at ``disnake.abc.Messageable.pins`` (defined in [disnake/abc.py](disnake/abc.py)).
+
+```py
+    async def pins(self) -> List[Message]:
+        channel = await self._get_channel()
+        state = self._state
+        data = await state.http.pins_from(channel.id)
+        return [state.create_message(channel=channel, data=m) for m in data]
+```
+*docstring removed for breverity*
+
+Here we have several things occuring. First, we have typehinted the return type of this method to return a list of Messages. As disnake supports python 3.8, we must use typing imports instead of subscripting built-ins--hence the capital ``List``.
+
+The next interesting thing is `self._state`. The library uses a state-centric design, which passes the state around to most objects. This is what enables edit methods and most methods on objects. Every Discord model that makes requests uses that internal state and its http attribute to make requests to the Discord API. Each endpoint is processed and defined in [disnake/http.py](disnake/http.py). This is where all of the methods are defined--and its where `pins_from` is defined too!
+
+The source of `pins_from` is as follows:
```

Comment:
```suggestion
The next interesting thing is `self._state`. The library uses a state-centric design, which passes the state around to most objects, which is what enables edit methods and most methods on objects. Every Discord model that makes requests uses that internal state and its http attribute to make requests to the Discord API. Each endpoint is processed and defined in [disnake/http.py](disnake/http.py) — and it's where `http.pins_from` is defined too, which looks like this:
```
Not sure about the "edit methods and most methods on objects" phrasing here

Reply:
Yeah, also not sure about that phrasing. https://github.com/DisnakeDev/disnake/pull/601/commits/b287c7c5404b798ec2598c6ae38c207fe129fa7c

Reply:
Perhaps just shorten it to `The library uses a state-centric design, which means the state is passed around to most objects.`, the next sentence already goes into further detail.

Reply:
https://github.com/DisnakeDev/disnake/pull/601/commits/f0fb8b09e18f4e4b6118d1ef7c4e1feca004e197